File: packages/jekko/src/cli/cmd/tui/feature-plugins/jankurai/delta.ts
// Delta-vs-baseline math + glyph chooser for the audit-live panel. The Δ
// columns in the panel render `value (+/-N glyph)` and `glyph` depends on
// whether higher-or-lower is better for the metric.

export type DeltaDirection = "improving" | "worsening" | "neutral" | "unknown"
export type DeltaMetric = "score" | "findings" | "caps" | "hard" | "soft" | "level"

export type DeltaResult = {
  /** Current minus baseline. `undefined` when either side is missing. */
  delta: number | undefined
  direction: DeltaDirection
  /** Single glyph the panel paints next to the delta number. */
  glyph: string
}

export function delta(current: number | undefined, baseline: number | undefined, metric: DeltaMetric): DeltaResult {
  if (current === undefined || baseline === undefined) {
    return { delta: undefined, direction: "unknown", glyph: "—" }
  }
  const diff = current - baseline
  if (diff === 0) {
    return { delta: 0, direction: "neutral", glyph: "=" }
  }
  // For score, higher is better. For finding counters, lower is better.
  const lowerIsBetter = metric !== "score" && metric !== "level"
  const improving = lowerIsBetter ? diff < 0 : diff > 0
  const direction: DeltaDirection = improving ? "improving" : "worsening"
  const magnitude = Math.abs(diff)
  // Double-arrow glyph above a large delta — purely cosmetic; the renderer
  // doesn't care about magnitude, but the panel uses ≥10 (score) or ≥10
  // (finding count) as the threshold.
  const big = magnitude >= 10
  const glyph = improving ? (big ? "▲▲" : "▲") : big ? "▼▼" : "▼"
  return { delta: diff, direction, glyph }
}

/** Format the delta for the panel as `±N glyph`. */
export function formatDelta(result: DeltaResult): string {
  if (result.delta === undefined) return "—"
  if (result.delta === 0) return "= 0"
  const sign = result.delta > 0 ? "+" : ""
  return `${sign}${result.delta} ${result.glyph}`
}

File: crates/jekko-tui/src/feature_plugins/jankurai/delta.rs
//! Delta computation for score / findings / caps vs baseline.

use ratatui::style::Color;

use super::style::{BLUE, GOLD, GREEN, RED};

/// Delta direction enum mirrors `delta.ts::DeltaDirection`.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum DeltaDirection {
    /// Metric improved vs baseline.
    Improving,
    /// Metric regressed vs baseline.
    Worsening,
    /// No change.
    Neutral,
    /// One side missing.
    Unknown,
}

impl DeltaDirection {
    pub(super) fn color(self) -> Color {
        match self {
            DeltaDirection::Improving => GREEN,
            DeltaDirection::Worsening => RED,
            DeltaDirection::Neutral => BLUE,
            DeltaDirection::Unknown => GOLD,
        }
    }
}

/// Whether a metric is "lower-is-better" (findings, caps) or "higher-is-better"
/// (score, level).
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
#[allow(dead_code)]
pub enum DeltaMetric {
    /// Score is higher-is-better.
    Score,
    /// Findings counter — lower is better.
    Findings,
    /// Caps applied — lower is better.
    Caps,
    /// Hard findings — lower is better.
    Hard,
    /// Soft findings — lower is better.
    Soft,
    /// Conformance level — higher is better.
    Level,
}

impl DeltaMetric {
    fn lower_is_better(self) -> bool {
        !matches!(self, DeltaMetric::Score | DeltaMetric::Level)
    }
}

/// One delta computation result. Ports the `DeltaResult` struct in
/// `delta.ts`.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct DeltaResult {
    /// `current - baseline`. None when either side is missing.
    pub delta: Option<f64>,
    /// Improvement/regression direction.
    pub direction: DeltaDirection,
    /// Glyph the panel draws beside the number (`▲`, `▼`, `=`, `-`).
    pub glyph: &'static str,
}

/// Compute the delta between `current` and `baseline` for a given metric.
///
/// Ports `delta.ts::delta`.
pub fn compute_delta(
    current: Option<f64>,
    baseline: Option<f64>,
    metric: DeltaMetric,
) -> DeltaResult {
    match (current, baseline) {
        (Some(c), Some(b)) => {
            let diff = c - b;
            if diff == 0.0 {
                return DeltaResult {
                    delta: Some(0.0),
                    direction: DeltaDirection::Neutral,
                    glyph: "=",
                };
            }
            let improving = if metric.lower_is_better() {
                diff < 0.0
            } else {
                diff > 0.0
            };
            let big = diff.abs() >= 10.0;
            let glyph = match (improving, big) {
                (true, true) => "▲▲",
                (true, false) => "▲",
                (false, true) => "▼▼",
                (false, false) => "▼",
            };
            DeltaResult {
                delta: Some(diff),
                direction: if improving {
                    DeltaDirection::Improving
                } else {
                    DeltaDirection::Worsening
                },
                glyph,
            }
        }
        _ => DeltaResult {
            delta: None,
            direction: DeltaDirection::Unknown,
            glyph: "-",
        },
    }
}

/// Format a [`DeltaResult`] like `+3 ▲` or `= 0`. Ports `delta.ts::formatDelta`.
pub fn format_delta(result: &DeltaResult) -> String {
    match result.delta {
        None => "-".to_string(),
        Some(0.0) => "= 0".to_string(),
        Some(d) => {
            let sign = if d > 0.0 { "+" } else { "" };
            // Treat near-integer deltas as whole numbers for terseness.
            if (d - d.round()).abs() < 1e-9 {
                format!("{}{} {}", sign, d.round() as i64, result.glyph)
            } else {
                format!("{}{:.1} {}", sign, d, result.glyph)
            }
        }
    }
}
